internal/gcsx/inactive_timeout_reader.go

```diff
@@ -0,0 +1,245 @@
+	return NewInactiveTimeoutReaderWithClock(ctx, bucket, object, readHandle, byteRange, timeout, clock.RealClock{})
+}
+
+func NewInactiveTimeoutReaderWithClock(ctx context.Context, bucket gcs.Bucket, object *gcs.MinObject, readHandle []byte, byteRange gcs.ByteRange, timeout time.Duration, clock clock.Clock) (gcs.StorageReader, error) {
+	if timeout == 0 {
+		return nil, ErrZeroInactivityTimeout
+	}
+
+	itr := &inactiveTimeoutReader{
+		object:     object,
+		bucket:     bucket,
+		reqRange:   byteRange,
+		readHandle: readHandle,
+		mu:         locker.New("inactiveTimeoutReader: "+object.Name, func() {}),
+		isActive:   false,
+	}
+	itr.ctx, itr.cancel = context.WithCancel(ctx)
+
+	var err error
+	if itr.gcsReader, err = itr.createGCSReader(); err != nil {
+		return nil, err
+	}
+
+	// Start the background periodic routine.
+	go itr.monitor(clock, timeout)
+
+	return itr, nil
+}
+
+// createGCSReader is a helper method to create the underlined reader from itr.start + itr.seen offset.
+func (itr *inactiveTimeoutReader) createGCSReader() (gcs.StorageReader, error) {
+	reader, err := itr.bucket.NewReaderWithReadHandle(
+		itr.ctx,
+		&gcs.ReadObjectRequest{
+			Name:       itr.object.Name,
+			Generation: itr.object.Generation,
+			Range: &gcs.ByteRange{
+				Start: itr.reqRange.Start + itr.seen,
+				Limit: itr.reqRange.Limit,
+			},
+			ReadCompressed: itr.object.HasContentEncodingGzip(),
+			ReadHandle:     itr.readHandle,
+		})
+	if err != nil {
+		return nil, fmt.Errorf("NewReaderWithReadHandle: %w", err)
+	}
+	return reader, nil
+}
+
+// Read implements io.Reader interface.
+//
+// If the underlying reader has been closed due to inactivity, Read automatically
+// attempts to reconnect using the stored read handle and the correct offset
+// (start + bytes previously seen). If reconnection fails, the error is returned.
+//
+// Each successful Read call marks the reader as active, resetting the inactivity timer
+// in the background monitor. This method is thread-safe.
+//
+// Calling Read() after explicitly calling Close() is not supported and will
+// lead to undefined behavior.
+func (itr *inactiveTimeoutReader) Read(p []byte) (n int, err error) {
+	itr.mu.Lock()
+	defer itr.mu.Unlock()
+
+	itr.isActive = true
+
+	if itr.gcsReader == nil {
+		if itr.gcsReader, err = itr.createGCSReader(); err != nil {
+			return 0, err
+		}
+	}
+
+	n, err = itr.gcsReader.Read(p)
+	itr.seen += uint64(n)
+	return n, err
+}
+
+// Close explicitly closes the underlying gcs.StorageReader if it's currently open.
+// It also signals the background monitor goroutine to stop.
+// Returns an error if closing the underlying reader fails.
+func (itr *inactiveTimeoutReader) Close() (err error) {
+	itr.mu.Lock()
+	defer itr.mu.Unlock()
+
+	itr.cancel = nil
+	itr.ctx = nil
+
+	if itr.gcsReader == nil {
+		return nil
+	}
+
+	err = itr.gcsReader.Close()
+	itr.gcsReader = nil
+	if err != nil {
+		return fmt.Errorf("close reader: %w", err)
+	}
+	return err
```

Comment:
nit: return nil?

Reply:
will take as part of next PR, concurrent-unit test PR.